### PR TITLE
amaranth{,-soc,-boards}: update and rename from nmigen, glasgow: unstable-2021-03-02 -> unstable-2021-12-12

### PR DESCRIPTION
**Pull request by leo60228**
###### Motivation for this change
https://twitter.com/whitequark/status/1477494676785709061

This is partially breaking: `amaranth` contains an `nmigen` module for backwards compatibility, but setuptools is unhappy about the change.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
  - manual inspection: glasgow likely needs an update
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by leo60228** [comment]
> manual inspection: glasgow likely needs an update

Well, this sure was the case... I guess I'll update it in this PR too?

**Comment by whitequark** [comment]
yeah, Glasgow definitely needs to be updated

**Comment by whitequark** [comment]
This PR will resolve the issue https://github.com/NixOS/nixpkgs/issues/140015.

File: pkgs/top-level/python-aliases.nix [hunk 1]
```diff
@@ -62,6 +62,9 @@ mapAliases ({
   Keras = keras; # added 2021-11-25
   lammps-cython = throw "lammps-cython no longer builds and is unmaintained"; # added 2021-07-04
   MechanicalSoup = mechanicalsoup; # added 2021-06-01
+  nmigen-boards = amaranth-boards; # added 2022-01-02
+  nmigen = amaranth; # added 2022-01-02
+  nmigen-soc = amaranth-soc; # added 2022-01-02
```
Comment (sbourdeauducq) [inline pkgs/top-level/python-aliases.nix:67]: nMigen is our registered trademark and those aliases arguably constitute infringement.

**Comment by sbourdeauducq** [comment]
While I welcome the renaming of the nMigen fork, please remove the aliases.

**Comment by Artturin** [comment]
make the aliases into throw's that point people to amaranth

**Comment by sbourdeauducq** [comment]
> make the aliases into throw's that point people to amaranth

That's precisely the trademark confusion that ought to be avoided. nMigen is a registered trademark (USPTO reg. no 6,318,709) and creating aliases under the nmigen name to another similar program (especially one that continues to make further confusing claims) is not OK. @edolstra

**Comment by videogame-hacker** [comment]
I think `throw` is a good solution, here. `nmigen` is not available on the system, but the packaging system knows about something that functions well in its place. Why shouldn't it help the user?

**Comment by sbourdeauducq** [comment]
Ask your lawyer.

**Review by emilazy — APPROVED** [review]
Seems fine to me; I can't imagine any realistic trademark issue from the wording of the `throw`. Even apart from my limited time and energy to devote to NixOS as of late, I don't have too much desire to engage with this package specifically for what I think are obvious reasons to anyone reading the past several PRs and issues relating to it at this point, but I'm happy to remain in the maintainers list especially with nobody else volunteering and I'll still try to review PRs when I have the time.

While I don't wish for this to be something the project has to think about in the first place, if it continues to come up in the future I think it would make sense for NixOS to adopt a stance on legal threats [akin to Wikipedia's](https://en.wikipedia.org/wiki/Wikipedia:No_legal_threats) to avoid chilling effects and the consequent poisoning of discourse.

**Comment by sbourdeauducq** [comment]
@emilazy stop playing the victim. Besides, the Wikipedia page you linked to even says "A complaint in cases of copyright infringement is not a legal threat", one could imagine the same would apply to complaints about trademark infringement.

**Comment by L-as** [comment]
So what's the motivation for this? The tweet is gone.

**Comment by sbourdeauducq** [comment]
> Ask your lawyer.

This is not a legal threat either, I am simply not in a position to educate you about trademark law. It means just that: you should ask your lawyer to explain you the answer to "why not do that".

**Comment by whitequark** [comment]
> So what's the motivation for this? The tweet is gone.

@L-as I asked someone familiar with nixpkgs to submit a PR renaming the packages. Once @leo60228 stepped up to do so, I've removed the tweet to avoid others doing duplicate work; I didn't realize it was cited in the "motivation" section.

**Comment by L-as** [comment]
In our case, we care about American law since GitHub is hosted there, and there were to be a violation, it would be the GitHub repository that will be removed or similar.
Doing `nmigen = amaranth` isn't clearly "fair use" wrt. American trademark law, and I'm leaning on the side that it's probably not legal, however, let's assume it's illegal, what is the trademark owner going to do? Sue the NixOS foundation? They don't make this change, nor do they own the code, They don't even host it! Sue GitHub? Very unlikely.

I'm not recommending that you do the change FWIW, in either case it's of very little importance really.

Some people have told me that packaging Amaranth is somehow also a trademark violation, but that's probably not the case, even if the Amaranth project is somehow violating trademark law. The code contained here merely refers to Amaranth.

IANAL of course, but law isn't as incomprehensible as most people like to imagine (especially EU law is easy to read IMO).
In any case, anyone suing anyone over this tiny change in Nixpkgs is extremely unlikely. If it does happen, we can just handle it then.

I don't see the issue with merging this personally.

**Review by L-as — APPROVED** [review]
LGTM

**Comment by L-as** [comment]
FWIW I'm not sure that *removing* nMigen is a good idea, since https://gitlab.com/nmigen/nmigen/-/commits/master/ seems to be in active development.

**Comment by zhaofengli** [comment]
> I think `throw` is a good solution, here. `nmigen` is not available on the system, but the packaging system knows about something that functions well in its place. Why shouldn't it help the user?
> 
> It's not automatically masquerading one package as another, it's simply informing the user that an alternative exists.

I agree. It's important to note that we are not "creating aliases" - It's just an informational message and manual configuration changes will be required. Even with aliases, there have been prior arts in other distributions where aliases to compatible or technically equivalent packages are added. Examples include Debian's  `mysql` -> `mariadb` (compatible), and historically `firefox` -> `iceweasel` (equivalent).

Either way, it doesn't really matter - I don't see us carrying M-Labs' version of nMigen, with or without the throw message.

**Comment by sbourdeauducq** [comment]
Iceweasel is not making confusing statements about Firefox though.

**Review by L-as — COMMENTED** [review]
LibreSOC will likely continue to use M-Labs's nmigen, as it contains changes not in amaranth AFAIK. It would be nice for me if you instead of removing nmigen, leave it as-is (with me as maintainer possibly, maintainers.l-as). I don't feel strongly against using amaranth as a dependency for glasgow etc.

**Comment by Artturin** [comment]
> LibreSOC will likely continue to use M-Labs's nmigen, as it contains changes not in amaranth AFAIK. It would be nice for me if you instead of removing nmigen, leave it as-is (with me as maintainer possibly, maintainers.l-as). I don't feel strongly against using amaranth as a dependency for glasgow etc.

we have never had M-Labs nmigen packaged. it has always been amaranth

**Comment by L-as** [comment]
You are in fact right.

**Comment by thoughtpolice** [comment]
I have pushed four out of the five commits on this branch directly to `master`, but I intentionally have not pushed the commit containing the `throw` aliases to `all-packages.nix` for nmigen, AKA e5fb34c7dffa60e07dcc3fc869da9193fef076a5. No aliases will be added, either. It does mean that downstream users will get an evaluation error rather than a message. That's just life sometimes. To be clear this is directly in line with the wishes expressed in [this comment](https://github.com/NixOS/nixpkgs/pull/153163#issuecomment-1008716691). The author has made it *repeatedly* clear here and elsewhere that they don't want these mixed up and they believe the name "nmigen" to be under their control, and has repeatedly made allusions to the situation surrounding it being a dangerous legal quagmire of some sorts with no limit to who they will drag into it. It is not arbitrary; I'm merely granting the authors' literal wish.

I see no reason to entertain any assumptions like "Well, I'm not a lawyer, but IMO I think it's fine" when I have no reason to believe there's any amicable solution that won't result in more shitposting here down the line; it would have already occurred. More generally when people conveniently drop by to make vague allusions to lawyers over the wordsmithing of an error message 15 people on earth might see &mdash; I'm not actually inclined to believe we share any side. There's no reason to prostrate or sit around and speculate, all to appease some dispute with terms only someone else is privy to, when the answer is quite simple: just wash our hands of it. Is that annoying for some downstream consumers? Maybe. Is it better for the health of Nixpkgs and everyone involved? Yes, and that is more valuable. In fact I'm actually more inclined to ensure `nmigen` never returns to the upstream package set again, in any form whatsoever, considering my only experience with its one-and-only owner has literally always been a waste of my time.

(And before anyone asks me about it, what useful error message can we even write if any mention of amaranth is met with "ask your lawyer" or some future complaints in another comment on another bug *again*? What, write `throw = "nmigen has been removed, there's an alternative package that literally everyone has migrated to you can use but that we can't name for reasons, maybe try google?";` You might as well just write `throw "Went to the farm";` and save yourself 25 words or whatever, but it's not even worth it.)

Given all this, I consider this issue and #140015 now closed as a result of this, and I see no reason to take any further input on the matter. And I'm more than happy to take responsibility for any fallout from that decision, and either clean it up or address it directly &mdash; as usual. Anyone suffering e.g. evaluation errors from this decision can either contact me or be referred to this ticket in the future; the solutions are all relatively straightforward (pin your nixpkgs version, write an expression yourself, or migrate to amaranth.)

---

As a side note, I have also added myself as a maintainer to the glasgow and amaranth packages. As Emily herself noted she isn't gone but mostly dormant; my own pace has slowed down but I do have a keen interest on keeping all the EDA tools up to date, this is just another set and otherwise has nobody to maintain it. I'll also have a Glasgow coming one of these days, so I actually will be a direct user anyway...

**Pull request closed without merging**